### PR TITLE
feat(python): enable "inefficient apply" warnings from `Series`

### PR DESCRIPTION
**Pull request by alexander-beedie**
Ref: #9968.

* Extends warnings about inefficient `Expr.apply` to `Series.apply`.
* Adds parsing recognition for additional builtin: `abs`.

## Example

```python
import polars as pl

srs = pl.Series( 
    name = "s", 
    values = [-20, -12, -5, 0, 5, 12, 20] 
).apply( 
    lambda x: (abs(x) != 12) and (x > 10 or x < -10 or x == 0) 
)

# PolarsInefficientApplyWarning: 
# Series.apply is significantly slower than the native series API.
# Only use if you absolutely CANNOT implement your logic otherwise.
# In this case, you can replace your `apply` with the following:
#   -  s.apply(lambda x: ...)
#   +  (s.abs() != 12) & ((s > 10) | ((s < -10) | (s == 0)))
```

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
nice - is it possible to add a test case using `abs(` to the `TEST_CASES` constant?

and if this addresses the issue brought up in 10103, do you want to cherry-pick e4eb9e8b065cbd05c665b69de59db61fd2b30910 ?

**Comment by alexander-beedie** [comment]
> nice - is it possible to add a test case using `abs(` to the `TEST_CASES` constant?
> 
> and if this addresses the issue brought up in 10103, do you want to cherry-pick [e4eb9e8](https://github.com/pola-rs/polars/commit/e4eb9e8b065cbd05c665b69de59db61fd2b30910) ?

Sounds like a plan! Let me have a look at doing exactly that... 👍

File: py-polars/polars/utils/udfs.py [hunk 1]
```diff
@@ -229,7 +231,7 @@ def rewritten_instructions(self) -> list[Instruction]:
         """The rewritten bytecode instructions from the function we are parsing."""
         return list(self._rewritten_instructions)
 
-    def to_expression(self, col: str, as_repr: bool = True) -> str | None:
+    def to_expression(self, col: str) -> str | None:
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:234]: RIP `as_repr`, 2023-07-26 - 2023-07-27

File: py-polars/tests/test_udfs.py [hunk 2]
```diff
@@ -65,6 +65,11 @@
     ("a", lambda x: MY_CONSTANT + x, 'MY_CONSTANT + pl.col("a")'),
     ("a", lambda x: 0 + numpy.cbrt(x), '0 + pl.col("a").cbrt()'),
     ("a", lambda x: np.sin(x) + 1, 'pl.col("a").sin() + 1'),
+    (
+        "a",  # note: functions operate on consts
+        lambda x: np.sin(3.14159265358979) + (x - 1) + abs(-3),
+        '(np.sin(3.14159265358979) + (pl.col("a") - 1)) + abs(-3)',
+    ),
```
Comment (MarcoGorelli) [inline py-polars/tests/test_udfs.py:72]: could we add a test case with `abs(x)` to `TEST_CASES` too please?

File: py-polars/polars/utils/udfs.py [hunk 3]
```diff
@@ -260,14 +262,16 @@ def to_expression(self, col: str, as_repr: bool = True) -> str | None:
...
+        polars_expr = " ".join(expr for _offset, expr in expression_strings)
 
         # note: if no 'pl.col' in the expression, it likely represents a compound
         # constant value (e.g. `lambda x: CONST + 123`), so we don't want to warn
         if "pl.col(" not in polars_expr:
             return None
-
-        return polars_expr if as_repr else eval(polars_expr, globals())
+        elif self._apply_target == "series":
+            return polars_expr.replace(f'pl.col("{col}")', "s")
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:272]: what if there's a global variable called `s`? might this conflict with that?

this is the case I'm worried about:
```python
In [5]: s = 3

In [6]: df = pl.DataFrame({'a': [1,2,3]})

In [7]: df['a'].apply(lambda x: x+s)
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for updating, just got some small comments

File: py-polars/polars/utils/udfs.py (hunk 3 above)
Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py:272]: Well played, lol... Yes, that would be an unlikely but real issue! I'll see about mitigating it, hmm 🤔

File: py-polars/tests/test_udfs.py (hunk 2 above)
Comment (alexander-beedie) [reply py-polars/tests/test_udfs.py:72]: Done 👌

File: py-polars/polars/utils/udfs.py [hunk 4]
```diff
@@ -178,6 +182,35 @@ def _inject_nesting(
...
+        else:
+            col_expr = f'pl.col("{col}")'
+            if self._apply_target == "expr":
+                return col_expr
+            elif self._apply_target == "series":
+                # handle overlapping name from global variables; fallback through
+                # "s", "srs", "series", and then the series' name (if applicable)
+                series_aliases = ["s", "srs", "series"]
+                if col.isalpha() and col not in series_aliases:
+                    series_aliases.append(col)
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:198]: why is this necessary?

File: py-polars/polars/utils/udfs.py [hunk 5]
```diff
@@ -178,6 +182,35 @@ def _inject_nesting(
...
+                for name in series_aliases:
+                    if not re.search(rf"\b{name}\b", search_expr):
+                        self._apply_target_name = name
+                        return name
+
+                # ...before, finally, srs0 -> srsN until we find an available name
+                n = count()
+                while True:
+                    name = f"srs{next(n)}"
+                    if not re.search(rf"\bs{name}\b", search_expr):
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:209]: why is there an extra 's' here? why not just 
```suggestion
                    if not re.search(rf"\b{name}\b", search_expr):
```
?

Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py]: Typo :)

File: py-polars/polars/utils/udfs.py (hunk 4 above)
Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py:198]: Should actually be `.isidentifier` rather than `isalpha` (I keep forgetting that exists - will update).

Comment (MarcoGorelli) [reply py-polars/polars/utils/udfs.py:198]: even so, not sure I sure why it's necessary at all. I tried deleting it and the tests passed - any chance you could expand on this please (or, even better, add a test for which it's necessary)?

Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py:198]: It's necessary if the series name has a space, starts with a number, or contains non-alphanumeric chars ;)

Comment (MarcoGorelli) [reply py-polars/polars/utils/udfs.py:198]: ah, right!
```python
In [3]: pl.Series('bob', [1,2,3]).apply(lambda x: x+s+srs+series)
<ipython-input-3-e199d2d5c974>:1: PolarsInefficientApplyWarning: 
Series.apply is significantly slower than the native series API.
Only use if you absolutely CANNOT implement your logic otherwise.
In this case, you can replace your `apply` with the following:
  -  bob.apply(lambda x: ...)
  +  ((bob + s) + srs) + series
```

How about we just delete this and go straight to `srs0`? it's already quite a lot to parse while reading :) 

final comment here, will approve after this one 😇

Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py:198]: > How about we just delete this and go straight to `srs0`? it's already quite a lot to parse while reading :)

Done; the chance of ever hitting it is already negligible - let's simplify!

File: py-polars/polars/utils/udfs.py [hunk 6]
```diff
@@ -178,6 +182,30 @@ def _inject_nesting(
...
+                search_expr = expression.replace(col_expr, "")
+                for name in ("s", "srs", "series"):
+                    if not re.search(rf"\b{name}\b", search_expr):
+                        self._apply_target_name = name
+                        return name
+                n = count()
+                while True:
+                    name = f"srs{next(n)}"
+                    if not re.search(rf"\b{name}\b", search_expr):
+                        return name
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:205]: so sorry 😢 
```suggestion
                        self._apply_target_name = name
                        return name
```

**Review by MarcoGorelli — APPROVED** [review]
Awesome, thanks for bearing with me 🙌

**Comment by alexander-beedie** [comment]
> Awesome, thanks for bearing with me 🙌

Lol... making the world a better place, one line of code at a time ;)
